### PR TITLE
WIP - use logger from ctx instead of scope

### LINKED ISSUE #1840: controller/openstackcluster: reconcileID is re-used 1000+ times
/kind bug

**What steps did you take and what happened:**

Open https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1789/pull-cluster-api-provider-openstack-e2e-test/1736818285917245440/artifacts/clusters/bootstrap/logs/capo-system/capo-controller-manager/capo-controller-manager-66896777f9-z6k2q/manager.log

And `grep` for:
```
reconcileID="12187d68-f3cf-4596-94e8-83666306d5fc"
```
2191 uses of the same reconcile ID.

That's wrong and it wasn't the case in late October:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-openstack/1696/pull-cluster-api-provider-openstack-e2e-test/1717592179636965376/artifacts/clusters/bootstrap/logs/capo-system/capo-controller-manager/capo-controller-manager-66896777f9-kfgvg/manager.log

I strongly suspect this [PR](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1688) to be the cause, I've bisected CI jobs and this issue started after this PR.

**What did you expect to happen:**

The reconcile loops should be uniquely used and IDs shouldn't be the same. I found this issue when debugging https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1788.

### PR DESCRIPTION
**Pull request by EmilienM**
**What this PR does / why we need it**:

Wwe create a cached provider scope and we put the logger in it. This has caused issue #1840 so now we'll get the logger from context.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1840

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Comment by EmilienM** [comment]
Before doing other files, I want to see the impact on the openstackcluster controller logs and if that changed anything.

File: controllers/openstackcluster_controller.go [hunk 1]
```diff
@@ -474,15 +477,17 @@ func bastionHashHasChanged(computeHash string, clusterAnnotations map[string]str
...
+
 	clusterName := fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)
 
 	networkingService, err := networking.NewService(scope)
 	if err != nil {
 		return err
 	}
 
-	scope.Logger().Info("Reconciling network components")
+	log.Info("Reconciling network components")
```
Comment (EmilienM) [inline controllers/openstackcluster_controller.go:490]: This seems to have worked:

* without that patch:
```
"Reconciling network components" controller="openstackcluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="OpenStackCluster" OpenStackCluster="e2e-b35o3t/cluster-e2e-b35o3t" namespace="e2e-b35o3t" name="cluster-e2e-b35o3t" reconcileID="1b3f9997-07dc-43e2-8e9b-f5d639bd4390" cluster="cluster-e2e-b35o3t"
```
count of reconcileID: 1000+

* with that patch:
```
"Reconciling network components" controller="openstackcluster" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="OpenStackCluster" OpenStackCluster="e2e-gik87x/cluster-e2e-gik87x" namespace="e2e-gik87x" name="cluster-e2e-gik87x" reconcileID="b35030d5-7612-46af-aac2-7473b5e35041"
```
count of reconcileID: 5

@mdbooth just confirming with you that my solution is ok with you and I'll replace all `scope.Logger()` by a `log` from `ctx`.

Thanks

Comment (mdbooth) [reply controllers/openstackcluster_controller.go:490]: This looks great!

File: controllers/openstackcluster_controller.go [hunk 2]
```diff
@@ -491,7 +496,7 @@ func reconcileNetworkComponents(scope scope.Scope, cluster *clusterv1.Cluster, o
 	}
 
 	if openStackCluster.Spec.NodeCIDR == "" {
-		scope.Logger().V(4).Info("No need to reconcile network, searching network and subnet instead")
+		log.Info("No need to reconcile network, searching network and subnet instead")
```
Comment (mdbooth) [inline controllers/openstackcluster_controller.go:499]: Caution when you're making all these mechanical changes: please don't lose the `.V(n)` bit.

```suggestion
		log.V(4).Info("No need to reconcile network, searching network and subnet instead")
```

**Review by mdbooth — COMMENTED** [review]
This looks great, and has the added advantage of starting to proliferate ctx most places.

Just the note about not losing explicit log levels when we make this change.

File: controllers/openstackcluster_controller.go (hunk 2 above)
Comment (EmilienM) [reply controllers/openstackcluster_controller.go:499]: oh right, thanks.

**Review by huxcrux — APPROVED** [review]
/lgtm

**Comment by mdbooth** [comment]
@EmilienM you mentioned you weren't happy with how invasive this change became. Based on that feedback I created https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1913, which is less invasive because it doesn't require changing the Logger() calls. Do you want to close this PR?

I feel like we're likely to end up touching all of this stuff anyway, e.g. to pass Context everywhere in Gophercloud v2. I also think this highlights a design smell in Scope (logger probably shouldn't be in there). Still, I'll be glad not to do that work today.

**Pull request closed without merging**